nautobot/core/api/serializers.py

```diff
@@ -315,3 +399,195 @@ class BulkOperationIntegerIDSerializer(serializers.Serializer):
 class GraphQLAPISerializer(serializers.Serializer):
     query = serializers.CharField(required=True, help_text="GraphQL query")
     variables = serializers.JSONField(required=False, help_text="Variables in JSON Format")
+
+
+class CustomFieldModelSerializerMixin(ValidatedModelSerializer):
+    """
+    Extends ModelSerializer to render any CustomFields and their values associated with an object.
+    """
+
+    computed_fields = SerializerMethodField(read_only=True)
+    custom_fields = CustomFieldsDataField(
+        source="_custom_field_data",
+        default=CreateOnlyDefault(CustomFieldDefaultValues()),
+    )
+
+    @extend_schema_field(OpenApiTypes.OBJECT)
+    def get_computed_fields(self, obj):
+        return obj.get_computed_fields()
+
+    def get_field_names(self, declared_fields, info):
+        """Ensure that "custom_fields" and "computed_fields" are always included appropriately."""
+        fields = list(super().get_field_names(declared_fields, info))
+        self.extend_field_names(fields, "custom_fields")
+        self.extend_field_names(fields, "computed_fields", opt_in_only=True)
+        return fields
+
+
+# TODO: remove in 2.2
+@class_deprecated_in_favor_of(CustomFieldModelSerializerMixin)
+class CustomFieldModelSerializer(CustomFieldModelSerializerMixin):
+    pass
+
+
+class RelationshipModelSerializerMixin(ValidatedModelSerializer):
```

Comment:
nit: move to `nautobot.core.api.mixins`?

Reply:
I did try that before, however `ValidatedModelSerializer` and `BaseModelSerializer` base classes for these mixins seem to create another circular import problem🫠